uw_oidc/middleware.py

```diff
@@ -0,0 +1,75 @@
+from django.contrib import auth
+from django.core.exceptions import ImproperlyConfigured
+from django.http import HttpResponse
+from uw_oidc.id_token import username_from_token
+from uw_oidc.exceptions import InvalidTokenError
+
+
+class IDTokenAuthenticationMiddleware:
+    """
+    Supports ID Token (issued by UW OIDC provider)
+    based request authentication for specified clients.
+    """
+    TOKEN_SESSION_KEY = 'uw_oidc_idtoken'
+
+    def __init__(self, get_response=None):
+        self.get_response = get_response
+
+    def process_view(self, request, view_func, view_args, view_kwargs):
+        if not hasattr(request, 'session'):
+            raise ImproperlyConfigured(
+                'This authentication middleware requires session middleware '
+                'to be installed. Edit your MIDDLEWARE setting to insert '
+                '"django.contrib.sessions.middleware.SessionMiddleware" '
+                'before "uw_oidc.middleware.IDTokenAuthenticationMiddleware".')
+
+        if 'HTTP_AUTHORIZATION' in request.META:
+            auth_token = request.META['HTTP_AUTHORIZATION']
+
+            if (request.user.is_authenticated and auth_token and
+                    auth_token == request.session.get(self.TOKEN_SESSION_KEY)):
```

Comment:
Token has not changed but token might be expired. So we may need to store expiration time in session and verify.